python/ray/tune/checkpoint_manager.py

```diff
@@ -198,7 +202,9 @@ def best_checkpoints(self):
     def _priority(self, checkpoint):
         result = flatten_dict(checkpoint.result)
         priority = result[self._checkpoint_score_attr]
-        return -priority if self._checkpoint_score_desc else priority
+        if self._checkpoint_score_desc:
+            priority = -priority
+        return (not is_nan(priority), priority, checkpoint.order)
```

Comment:
When `priority` is `nan`, the sort by tuple key:

```
(not is_nan(priority), priority, checkpoint.order)
```

won't give the correct order by `checkpoint.order`. Because both `nan < nan` and `nan > nan` return `False`.

```suggestion
        if self._checkpoint_score_desc:
            priority = -priority
        if is_nan(priority):
            return (0, checkpoint.order, priority)
        return (1, priority, checkpoint.order)
```

Reply:
Actually it does:

```
>>> import numpy as np
>>> (False, np.nan, 3) < (False, np.nan, 4)
True
>>> (False, np.nan, 4) < (False, np.nan, 3)
False
```

Reply:
> Actually it does:
> 
> ```
> >>> import numpy as np
> >>> (False, np.nan, 3) < (False, np.nan, 4)
> True
> >>> (False, np.nan, 4) < (False, np.nan, 3)
> False
> ```

Seams that `tuple.__lt__` skips items when `lhs[i] is rhs[i]`.

```python
In [1]: (False, float('nan'), 3) < (False, float('nan'), 4)
Out[1]: False

In [2]: (False, float('nan'), 3) > (False, float('nan'), 4)
Out[2]: False

In [3]: (False, float('nan'), 4) < (False, float('nan'), 3)
Out[3]: False

In [4]: (False, float('nan'), 4) > (False, float('nan'), 3)
Out[4]: False

In [5]: import numpy as np

In [6]: (False, np.nan, 3) < (False, np.nan, 4)
Out[6]: True

In [7]: (False, np.nan, 3) > (False, np.nan, 4)
Out[7]: False

In [8]: import math

In [9]: (False, math.nan, 3) < (False, math.nan, 4)
Out[9]: True

In [10]: (False, math.nan, 3) > (False, math.nan, 4)
Out[10]: False

In [11]: float('nan') is float('nan')
Out[11]: False

In [12]: np.nan is np.nan
Out[12]: True

In [13]: math.nan is math.nan
Out[13]: True

In [14]: float('nan') is math.nan
Out[14]: False

In [15]: math.nan is np.nan
Out[15]: False

In [16]: (False, math.nan, 3) < (False, np.nan, 4)
Out[16]: False
```

Reply:
Ah thanks, you're right. It seems indeed like `float('nan') <= float('nan')` is `False`, unlike for np.

Fix here: https://github.com/ray-project/ray/pull/23909